images/virtualization-artifact/pkg/controller/vmip/internal/state/state.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmip/internal/util"
+	"github.com/deckhouse/virtualization-controller/pkg/sdk/framework/helper"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+type VMIPState interface {
+	VirtualMachineIP() *virtv2.VirtualMachineIPAddress
+	VirtualMachineIPStatus() *virtv2.VirtualMachineIPAddressStatus
+	VirtualMachineIPLease(ctx context.Context) (*virtv2.VirtualMachineIPAddressLease, error)
+	VirtualMachine(ctx context.Context) (*virtv2.VirtualMachine, error)
+
+	AllocatedIPs() common.AllocatedIPs
+}
+
+type state struct {
+	client       client.Client
+	vmip         *virtv2.VirtualMachineIPAddress
+	vmipStatus   *virtv2.VirtualMachineIPAddressStatus
+	lease        *virtv2.VirtualMachineIPAddressLease
+	vm           *virtv2.VirtualMachine
+	allocatedIPs common.AllocatedIPs
+}
+
+func New(c client.Client, vmip *virtv2.VirtualMachineIPAddress, vmipStatus *virtv2.VirtualMachineIPAddressStatus) VMIPState {
+	return &state{client: c, vmip: vmip, vmipStatus: vmipStatus}
+}
+
+func (s *state) VirtualMachineIP() *virtv2.VirtualMachineIPAddress {
+	return s.vmip
+}
+
+func (s *state) VirtualMachineIPStatus() *virtv2.VirtualMachineIPAddressStatus {
+	return s.vmipStatus
+}
+
+func (s *state) VirtualMachineIPLease(ctx context.Context) (*virtv2.VirtualMachineIPAddressLease, error) {
+	if s.lease != nil {
+		return s.lease, nil
+	}
+
+	var err error
+
+	leaseName := common.IpToLeaseName(s.vmip.Status.Address)
+
+	if leaseName != "" {
+		leaseKey := types.NamespacedName{Name: leaseName}
+		s.lease, err = helper.FetchObject(ctx, leaseKey, s.client, &virtv2.VirtualMachineIPAddressLease{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to get Lease %s: %w", leaseKey, err)
+		}
+	}
+
+	if s.lease == nil {
+		s.allocatedIPs, err = util.GetAllocatedIPs(ctx, s.client)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return s.lease, nil
+}
+
+func (s *state) VirtualMachine(ctx context.Context) (*virtv2.VirtualMachine, error) {
+	if s.vm != nil {
+		return s.vm, nil
+	}
+
+	var err error
+	if s.vmip.Status.VirtualMachine != "" {
+		vmKey := types.NamespacedName{Name: s.vmip.Status.VirtualMachine, Namespace: s.vmip.Namespace}
+		s.vm, err = helper.FetchObject(ctx, vmKey, s.client, &virtv2.VirtualMachine{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to get VM %s: %w", vmKey, err)
+		}
+	}
+
+	if s.vm == nil {
+		var vms virtv2.VirtualMachineList
+		err = s.client.List(ctx, &vms, &client.ListOptions{
+			Namespace: s.vmip.Namespace,
+		})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		vmipNameWithoutHash := s.vmip.Name
+		vmipNameWithoutHash = vmipNameWithoutHash[:len(vmipNameWithoutHash)-6]
+		for _, vm := range vms.Items {
+			if vm.Spec.VirtualMachineIPAddress == s.vmip.Name ||
+				vm.Spec.VirtualMachineIPAddress == "" && vm.Name == vmipNameWithoutHash {
+				s.vm = new(virtv2.VirtualMachine)
+				*s.vm = vm
+				break
```

Comment:
List() returns DeepCopied objects. Why second copy? Simple assign is enough here.

Reply:
Done